doc/adr/0003-directory-structure.md

```diff
@@ -0,0 +1,88 @@
+# 3. Define directory structure
+
+Date: 2020-12-01
+
+## Status
+
+Accepted
+
+## Context
+
+There will be many services contributed by multiple people. We need a common way of structuring the services
+so it's easy for the user to pick up any one service.
+
+Some services might want to share some common code. We want to have that code structured so that include path shouldn't
+change if these extensions are included into Mbed OS.
+
+## Decision
+
+### Services
+
+Each service will be its own library. All services will be in the services directory.
+Each service will contain at least an include folder and a mbed_lib.json.
+
+An example service called `example` would look like this:
+
+```
+services/example/include/ble-service-example/example.h
+services/example/source/example.cpp
+services/example/mbed_lib.json
+```
+
+and the mbed_lib.json would contain at least:
+
+```
+{
+    "name": "ble-service-example"
+}
+```
+
+This way a user can clone the repo and include an individual service without getting unwanted includes.
+
```

Comment:
Not to make this PR take longer to integrate, but what about Client code for a given service?

I'm thinking it should share the same root directory so that headers can be shared between the Service and Client implementations.

For example, I can foresee a `CurrentTimeService` and a corresponding `CurrentTimeServiceClient` (or should it just be `CurrentTimeClient`?)

It would be nice to be able to share definitions (such as UUIDs, etc) while having a separate `mbed_lib.json`. The include and source code for the Client and Service can share the same directories, but I'm wondering what you guys think.

The Client specification should be placed between the Service and Extensions section.

Reply:
Perhaps we could treat clients similar to extensions and put them in a separate directory. The client can then "require" the inclusion of the corresponding service.

Does this sound like a reasonable compromise? This would prevent having to pull in client code when the user only wants the service itself, while also allowing the client to share code with the service.

Reply:
Maybe client code could just be put together with the service? there are #defines that include or exclude code based on whether gatt server or gatt client is present.

Reply:
But your solution sounds good to me too (with clients being their own lib requiring the service to get the headers). @pan-?

Reply:
I like both of your approach; only clients would require the compilation of the server code. Paul's approach has merit too. 

Maybe we can leave it open for now and define a new ADR when the case occur.

Reply:
That sounds alright to me. We can address this when the client issue comes up.